Feature/atr 8 dev no bql field

SENya1990 opened the pull request:
Added PX1065 diagnostic to check DAC field properties for missing BQL fields + refactored common logic out of  legacy BQL fields and constants analyzers.

**Overview:**

**PX1065 diagnostic:**
- Added PX1065 diagnostic to check DAC field properties for missing BQL fields
- Added a code fix for PX1065 to generate a BQL field if the reported property is declared inside the analyzed DAC. 
- Added unit tests for PX1065 and its code fix for DACs and DAC extensions
- Added tests for PX1065 with DACs and DAC extension that derives or extend DACs from the external DLL
- Wrote documentation for PX1065

**Test Infrastructure:**
- Added external solution to the repo to imitate external dependency. Tests now are referencing the external dependency DLL. The dynamically generated test workspace also references external dependency DLL. 
- Fixed bug in `CodeFixVerifier` that caused code fix verification to exit prematurely if the first diagnostic didn't have any code actions registered
- Added capability to override the number of attempts to apply code actions in `CodeFixVerifier`

**Semantic Model** 
- Moved common logic for mapping between DAC property type and BQL field type into a separate
 `PropertyTypeToBqlFieldTypeMapping` helper
- Moved common logic for BQL field generation into utils and reworked legacy BQL field and constant code fixes.
- Extended `DacFieldAttributeInfo` with `IsAcumaticaAttribute` flag + refactoring
- Extended `IOverridableItem` interface with `CombineWithBaseInfo` to add to existing semantic infos initial support for combination with their base members. The combination should calculate _effective_ properties of the semantic info object based on its declared properties and properties of its base info object. Implemented initial support for this combination.
- Extended DAC semantic model and renamed part of its properties to be more clear on terminology
- Extended DAC property and DAC field infos with `IsAcumaticaAttribute` flag
- Fixed big issue in the DAC field infos collection. The combination with the base DAC field info wasn't done previously.

**Other Code Analysis Changes**
- Reworked `MissingTypeListAttributeAnalyzer ` to support aggregator attributes and use precalculated data about them. PX1002 check became more precise. 
- Fixed mass code fixes for diagnostics:
    - PX1012
    - PX1062
    - PX1072

**Utilities Changes:**
- Refactoring - renamed string type name constants and symbols with misleading names
- Added read only hash set wrapper for hashset 
- Small fixes in helpers

**Other:**
- Updated existing diagnostic docs - replaced `IBqlTable` with `PXBqlTable`

SENya1990 committed:
ATR-8: added resource strings and descriptor for PX1065 diagnostic

SENya1990 committed:
ATR-8: added read only hash set collection that wraps a mutable hashset and works faster on average than the immutable hashset

SENya1990 committed:
ATR-8: slightly optimized attribute info + added IsAcumatica attribute property

SENya1990 committed:
ATR-8: renamed IsDacFieldProperty to HasBqlField

SENya1990 committed:
ATR-8: added flag indicating if property info has Acumatica attributes

SENya1990 committed:
ATR-8: added extension points for combination of semantic infos with their base infos

SENya1990 committed:
ATR-8: extended DAC semantic model + refactoring

SENya1990 committed:
ATR-8: reworked missing type list analyzer to use pre-calced info from semantic model for aggregated field attributes

SENya1990 committed:
ATR-8: extended DAC model

SENya1990 committed:
ATR-8: reworked DAC field info's IsDacProperties into two separate properties and added HasAcumaticaAttributes property

SENya1990 committed:
ATR-8: implemented analysis for DAC properties with missing BQL fields

SENya1990 committed:
ATR-8: made check for registered diagnostic case insensitive

SENya1990 committed:
ATR-8: added mappings between BQL field and DAC propety type

SENya1990 committed:
ATR-8: refactoring - made IsRegisteredForCodeFix go through the TryGetPropertyValueInternal private helper

SENya1990 committed:
ATR-8: reworked legacy BQL field code fix to use new PropertyTypeToBqlFieldTypeMapping type

SENya1990 committed:
ATR-8: refactoring of legacy BQL field code fix - added empty lines + renamed local variables

SENya1990 committed:
ATR-8: reworked legacy BQL constant code fix to use new PropertyTypeToBqlFieldTypeMapping type

SENya1990 committed:
ATR-8: refactored legacy BQL constant code fix - renamed variables + empty lines

SENya1990 committed:
ATR-8: refactoring - renamed IsDacField to IsDacBqlField to correct terminology

SENya1990 committed:
ATR-8: extended mappping helper between property type and BQL field type with Contains methods and intergrated changes.

Details:
- extended mappping helper between property type and BQL field type with Contains methods that check if type is present in mapping tables
- integrated changes into legacy BQL field and constant analyzers
- removed mapping table from the BQL field analyzer since it is moved to the mapping helper in utilities

SENya1990 committed:
ATR-8: renamed BQL data type symbols and name constants to make them same as the Acumatica types and to reflect that they are interfaces

SENya1990 committed:
ATR-8: moved to common DAC utils helper that checks if BQL field or BQL constant

SENya1990 committed:
ATR-8: added common constant diagnostic property name PropertyType to pass property type to the code fix

SENya1990 committed:
ATR-8: added code generation helper method to generate BQL field node

SENya1990 committed:
ATR-8: integrated new code generation helper into the legacy BQL field code fix

SENya1990 committed:
ATR-8: implemented code fix that will generate a missing BQL field for a DAC property

SENya1990 committed:
ATR-8: refactoring - renamed BQL fields tests

SENya1990 committed:
ATR-8: added unit tests for the diagnostic for missing BQL fields

SENya1990 committed:
ATR-8: added new analyzer to all aggregated DAC analyzers

SENya1990 committed:
ATR-8: fixed an incorrect name of a diagnostic property in the code fix

SENya1990 committed:
ATR-8: fixed name retrieval for a property type

SENya1990 committed:
ATR-8: enhanced code fix generation to make new BQL fields one liners

SENya1990 committed:
ATR-8: copied regions from property leading trivia to the BQL field trivia

SENya1990 committed:
ATR-8: fixed code fix to support mass code fixes

SENya1990 committed:
ATR-8: fixed batch processing in other code fixes

SENya1990 committed:
ATR-8: added source for code fix test for PX1065

SENya1990 committed:
ATR-8: added test for PX1065 for derived DAC with base DAC without a BQL field

SENya1990 committed:
ATR-8: fixed calculation of the location to report and code fix registration logic

SENya1990 committed:
ATR-8: added PX1065 diagnostic to documentation

SENya1990 committed:
ATR-8: replaced IBqlTable with PXBqlTable in Acuminator docs to update code examples

SENya1990 commented on docs/Summary.md:
```diff
@@ -72,6 +72,7 @@ Acuminator does not perform static analysis of projects whose names contain `Tes
 | [PX1062](diagnostics/PX1062.md) | You cannot declare static views, actions, mutable fields, and mutable properties in graphs or graph extensions  | Error | Available |
 | [PX1063](diagnostics/PX1063.md) | The declaration of the exception class does not contain a serialization constructor. This will cause an incorrect deserialization of the exception data and will lead to runtime errors in Acumatica ERP. | Error | Available |
 | [PX1064](diagnostics/PX1064.md) | The declaration of the exception class introduces new serializable fields but does not declare an override for the `GetObjectData` method. This will cause an incorrect serialization of the exception data and will lead to runtime errors in Acumatica ERP. | Error | Available |
+| [PX1065](diagnostics/PX1065.md) | The DAC field property does not have a corresponding DAC BQL field. | Error | Available |
```
@EAndrosova  please review the diagnostic summary docs

SENya1990 commented on docs/diagnostics/PX1065.md:
```diff
@@ -0,0 +1,80 @@
+# PX1065
+This document describes the PX1065 diagnostic.
+
```
@EAndrosova  please review the diagnostic documentation page

SENya1990 commented on src/Acuminator/Acuminator.Analyzers/DiagnosticsShortName.resx:
```diff
@@ -402,4 +402,7 @@
   <data name="PX1022Graph" xml:space="preserve">
     <value>NonPublicGraph</value>
   </data>
+  <data name="PX1065" xml:space="preserve">
```
@EAndrosova  please review the diagnostic short name

SENya1990 commented on src/Acuminator/Acuminator.Analyzers/Resources.resx:
```diff
@@ -702,4 +702,10 @@ public virtual int? SomeDacField{ get; set; }</value>
   <data name="PX1022GraphExtensionFix" xml:space="preserve">
     <value>Make the graph extension public</value>
   </data>
+  <data name="PX1065TitleFormat" xml:space="preserve">
```
@EAndrosova  please review the diagnostic and code fix messages

dropsonic commented on src/Acuminator/Acuminator.Tests/Tests/StaticAnalysis/NoBqlFieldForDacFieldProperty/Sources/DacWithBqlFieldMissingInBaseDac.cs:
What about adding a test for a property that does not have any `PXEventSubscriberAttribute`-inherited attribute and checking that the diagnostic is not reported on such property since it is not "DAC field"?

SENya1990 replied:
great idea! I'll add it

SENya1990 replied:
Actually, after I added new tests, I decided to also add a tests with base DAC in external dll, so I've integrated one into the repo and tests. And with it I discovered a huge bug in the way this version of Roslyn does collection of data from the external symbols. `ISymbol.GetTypeMembers` method returns type itself instead of nested types.

So, it was a great idea to add new tests.

SENya1990 committed:
ATR-8: fixed bug in the code fix verifier.

The verifier previously used only the first diagnostic in the set to get available code actions. If that diagnostic couldn't provide any code action to apply, then the verifier stopped. This led to a situation when the first diagnostic in the code source didn't provide an action, and the verifier stops even if the second diagnostic can provide valid code actions

SENya1990 committed:
ATR-8: added C# property without Acumatica attributes to the test source for regulsr DAC

SENya1990 committed:
ATR-8: added second code fix test + added test for C# property with no Acumatica attributes to the base DAC

SENya1990 committed:
ATR-8: added new internal solution for external dependencies to tests.

This solution should be used to test Acuminator data collection and analysis from the external dll

SENya1990 committed:
ATR-8: added external dependency DLL folder to .gitignore

SENya1990 committed:
ATR-8: added base DAC to external dependency

SENya1990 committed:
ATR-8: Added reference to the external dependency to Acuminator tests

SENya1990 committed:
ATR-8: added external dependnecy dll to Git

SENya1990 committed:
ATR-8: added reference to external dependency dll to Acuminator tests and to the project generated for tests

SENya1990 committed:
ATR-8: fixed external dependency .Net framework version

SENya1990 committed:
ATR-8: changed built external dependency dlls to debug ones

SENya1990 committed:
ATR-8: added new test for DAC with external dll

SENya1990 committed:
ATR-8: fixed DAC in the external dependency

SENya1990 committed:
ATR-8: enabled test for DAC derived from external DAC

SENya1990 committed:
ATR-8: extended diagnostic to work for DAC extensions

SENya1990 committed:
ATR-8: fixed bug in field info collection - reworked combination of DAC field with its base info

SENya1990 committed:
ATR-8: reworked code fix verifier to support custom attempts count

SENya1990 committed:
ATR-8: added tests for DAC extension with missing BQL fields

SENya1990 committed:
ATR-8: added test for DAC extension on external DAC with missing BQL field

EAndrosova committed:
ATR-8: Edit doc for diagnostic
